### ISSUE TITLE
Defining IOTSA_WITH_BLE isn't enough to enable HPS

### ISSUE THREAD

**Issue by jackjansen** [cwi-dis/iotsa#84]
Because it turns out that if the code doesn't declare 

```
IotsaBLEServerMod bleserverMod(application);
```

then there isn't any BLE service that HPS can be implemented on.

Should rethink how that works, possibly by auto-configuring iotsaBLEServerMod when needed.

**Comment by jackjansen** [comment]
Sequencing note: this is a concrete instance of the general problem #85 (unify initialisation of iotsa standard modules) is addressing — #85 explicitly cites this issue as an example. Landing #85's general mechanism first could make this close to free rather than a one-off guard-clause fix.

— Claude Code (Claude Sonnet 5)

**Comment by jackjansen** [comment]
Sequencing note (2026-08-25): soft dependency on #113 — this issue (auto-create
`IotsaBLEServerMod` when HPS needs one) touches the same BLE-server creation path
#113's advertising-ordering fix touches, so it's cheaper done after #113 lands rather
than working around the ordering bug now. Also soft-generalized by #85 (unify
initialisation of standard modules) — landing #85 first could make this near-free
instead of a one-off fix. Full reasoning on cwi-dis/iotsa#199.

— Claude Code (Claude Sonnet 5)

**Comment by jackjansen** [comment]
Concrete motivation for this, found while working #205 (build-flag naming): once BLE client becomes first-class and BLE-client-only builds are real, `IOTSA_WITH_HPS`'s derivation (`#ifdef IOTSA_WITH_BLE`, soon `IOTSA_HAS_HPSSERVER`) becomes actively wrong for them, not just imprecise. HPS is inherently peripheral/server-role behavior (it publishes GATT characteristics for a remote central to read/write) -- so a client-only device would get HPS turned on for a server role it doesn't have, with nothing to catch it.

This is the same root cause already named here (nothing guarantees a BLE *server* object exists when `IOTSA_WITH_BLE` is on -- no `ensureServiceMod()`-style auto-creation for `IotsaBLEServerMod` the way CoAP/HPS's own companion mods get), just a sharper failure mode once client-only ships: today it's "assumes a server that might not exist," post-client-split it becomes "assumes a role the device was never going to have."

Not asking for scope change here -- just recording it so the motivation doesn't get lost by the time this is picked up. Also flagged as a parked item on #205, since fixing `IOTSA_HAS_HPSSERVER`'s derivation to depend on a real server-role signal is blocked on whatever this issue lands.

— Claude Code (Claude Sonnet 5)

**Comment by jackjansen** [comment]
Fixed on branch `85-standard-modules` (commit `8586b51`, depends on #85's `2b8f0ab`).

`IotsaApplication::setup()` now calls `IotsaBLEServerMod::ensure(*this)` under `IOTSA_WITH_BLE`, next to the existing CoAP/HPS `ensureServiceMod()` calls. So any BLE service (HPS, the battery service, an app's own) that registers against the single `IotsaBLEServerMod` now gets a module instance -- and therefore `lateSetupDone()` -> advertising -- without the sketch having to declare `IotsaBLEServerMod` explicitly. An explicit declaration still wins via the singleton.

Broader than the title: it's not just HPS, it's every BLE service. Gated on `IOTSA_WITH_BLE` for now; when BLE-client-only builds become real this moves to a server-role signal (per the discussion above).
Verified: ~12 PlatformIO envs green (esp8266/esp32/c3/s3, BLE+CoAP+HTTPS, WiFi-less), plus on hardware -- lolin32 running KitchenSink, full build (HTTP/HTTPS/HPS) and an `IOTSA_WITHOUT_WIFI` build (HPS + config still work).

Not merged to `develop` yet -- part of the layer-1a batch (see cwi-dis/iotsa#199).

-- Claude Code (Claude Sonnet 5)

**Comment by jackjansen** [comment]
Done -- `8586b51` is on `develop` (merged from `85-standard-modules`).

`IotsaApplication::setup()` now calls `IotsaBLEServerMod::ensure(*this)` under `IOTSA_WITH_BLE`, next to the existing CoAP/HPS `ensureServiceMod()` calls. Any BLE service (HPS, the battery service, an app's own) that registers against the single `IotsaBLEServerMod` now gets a module instance -- and therefore `lateSetupDone()` -> advertising -- without the sketch declaring `IotsaBLEServerMod` explicitly. An explicit declaration still wins via the singleton (see #85 for the mechanism).

Broader than the title in the end: it's every BLE service, not just HPS. Gated on `IOTSA_WITH_BLE` for now; when BLE-client-only builds become real this moves to a server-role signal.

Verified on hardware (lolin32 KitchenSink, HPS reachable) and across ~12 build envs.

-- Claude Code (Claude Sonnet 5)
